src/transformation/transformation.c

```diff
@@ -258,6 +310,148 @@ k4a_result_t transformation_2d_to_2d(const k4a_calibration_t *calibration,
+    {
+        // Compute the ray from the depth camera oringin, intersecting with the current searching pixel on the epipolar
+        // line
+        float ray[3];
+        ray[0] = (p[0] - pinhole.px) / pinhole.fx;
+        ray[1] = (p[1] - pinhole.py) / pinhole.fy;
+        ray[2] = 1.f;
+
+        // Project the ray to the distorted depth image to read the depth value from nearest pixel
+        float depth_point2d[2];
+        int p_valid = 0;
+        if (K4A_FAILED(TRACE_CALL(transformation_3d_to_2d(
+                calibration, ray, K4A_CALIBRATION_TYPE_DEPTH, K4A_CALIBRATION_TYPE_DEPTH, depth_point2d, &p_valid))))
+        {
+            return K4A_RESULT_FAILED;
+        }
+
+        if (p_valid == 1)
+        {
+            // Transform the current searching depth pixel to color image
+            if (transformation_is_pixel_within_image(depth_point2d,
+                                                     depth_image_width_pixels,
+                                                     depth_image_height_pixels))
+            {
+                int u = (int)(floorf(depth_point2d[0] + 0.5f));
+                int v = (int)(floorf(depth_point2d[1] + 0.5f));
+                uint16_t d = depth_image_data[v * depth_image_width_pixels + u];
+                float reprojected_point2d[2];
+                if (K4A_FAILED(TRACE_CALL(transformation_2d_to_2d(calibration,
+                                                                  depth_point2d,
+                                                                  d,
+                                                                  K4A_CALIBRATION_TYPE_DEPTH,
+                                                                  K4A_CALIBRATION_TYPE_COLOR,
+                                                                  reprojected_point2d,
+                                                                  &p_valid))))
+                {
+                    return K4A_RESULT_FAILED;
+                }
+
+                if (p_valid == 1)
+                {
+                    if (transformation_is_pixel_within_image(reprojected_point2d,
+                                                             calibration->color_camera_calibration.resolution_width,
+                                                             calibration->color_camera_calibration.resolution_height))
+                    {
+
+                        // Compute the 2d reprojection error and store the minimum one
+                        float error = sqrtf(powf(reprojected_point2d[0] - source_point2d[0], 2) +
+                                            powf(reprojected_point2d[1] - source_point2d[1], 2));
+                        if (error < best_error)
+                        {
+                            best_error = error;
+                            target_point2d[0] = depth_point2d[0];
+                            target_point2d[1] = depth_point2d[1];
+                            *valid = 1;
```

Comment:
>*valid = 1; [](start = 28, length = 11)

Note sure if we should return valid if the best_error is very large. What about occlusions?

Reply:
Will add a check on the best_error

---
In reply to: [319260972](https://github.com/microsoft/Azure-Kinect-Sensor-SDK/pull/696#discussion_r319260972) [](ancestors = 319260972)